args/core/containers/sparse_set.hpp

```diff
@@ -0,0 +1,269 @@
+#pragma once
+#include <vector>
+#include <type_traits>
+#include <algorithm>
+#include <core/platform/platform.hpp>
+
+/**
+ * @file sparse_set.hpp
+ */
+
+
+namespace args::core
+{
+	/**@class sparse_set
+	 * @brief Quick lookup contiguous set. A sparse set uses a dense and a sparse array to allow quick lookup of it's values whilst maintaining contiguous itteration.
+	 * @note Inserting items with large value differences does explode the sparse array and thus increases memory size.
+	 *       If you need support for items with large value differences such as hashes then please use the sparse_map.
+	 * @ref args::core::sparse_map
+	 * @tparam value_type The type to be used as the value (must be unsigned).
+	 * @tparam dense_type Container to be used to store the values.
+	 * @tparam sparse_type Container to be used to store the index mapping into dense container.
+	 * @note With default container parameters iterators may be invalidated upon resize. See reference of std::vector.
+	 * @note Removing item might invalidate the itterator of the last item in the dense container.
+	 */
+	template <typename value_type, template<typename...> typename dense_type = std::vector, template<typename...> typename sparse_type = std::vector>
+	class sparse_set
+	{
+		static_assert(std::is_unsigned_v<value_type>, "value_type must an unsigned type.");
+	public:
+		using size_type = std::size_t;
+
+		using sparse_container = sparse_type<value_type>;
+		using dense_container = dense_type<value_type>;
+
+		using reference = value_type&;
+		using const_reference = const value_type&;
+
+		using iterator = typename dense_container::iterator;
+		using const_iterator = typename dense_container::const_iterator;
+
+	private:
+		dense_container m_dense;
+		sparse_container m_sparse;
+
+		size_type m_size = 0;
+		size_type m_capacity = 0;
+
+	public:
+		A_NODISCARD inline iterator begin() { return m_dense.begin(); }
```

Comment:
the `inline`'s here are redundant and should be removed, the specifier is only required when actually implementing the function

Reply:
and or when not in a class definition